shortfin/tests/apps/llm/cpu_llm_server_test.py

```diff
@@ -0,0 +1,156 @@
+                    "-m",
+                    "sharktank.examples.export_paged_llm_v1",
+                    f"--gguf-file={model_path}",
+                    f"--output-mlir={mlir_path}",
+                    f"--output-config={config_path}",
+                    f"--bs={bs_string}",
+                ],
+                check=True,
+            )
+        # Compile model if it doesn't exist
+        vmfb_path = os.path.join(tmp_dir, "model.vmfb")
+        if not os.path.exists(vmfb_path):
+            subprocess.run(
+                [
+                    "iree-compile",
+                    mlir_path,
+                    "-o",
+                    vmfb_path,
+                ]
+                + settings["device_flags"],
+                check=True,
+            )
+        # Write config if it doesn't exist
+        edited_config_path = os.path.join(tmp_dir, "edited_config.json")
+        if not os.path.exists(edited_config_path):
+            config = {
+                "module_name": "module",
+                "module_abi_version": 1,
+                "max_seq_len": 2048,
+                "attn_head_count": 32,
+                "attn_head_dim": 100,
+                "prefill_batch_sizes": BATCH_SIZES,
+                "decode_batch_sizes": BATCH_SIZES,
+                "transformer_block_count": 26,
+                "paged_kv_cache": {"block_seq_stride": 16, "device_block_count": 256},
+            }
+            with open(edited_config_path, "w") as f:
+                json.dump(config, f)
+        yield tmp_dir
+    finally:
+        shutil.rmtree(tmp_dir)
+
+
+@pytest.fixture(scope="module")
+def llm_server(model_test_dir):
+    # Start the server
+    server_process = subprocess.Popen(
+        [
+            "python",
+            "-m",
+            "shortfin_apps.llm.server",
+            f"--tokenizer={os.path.join(model_test_dir, 'tokenizer.json')}",
+            f"--model_config={os.path.join(model_test_dir, 'edited_config.json')}",
+            f"--vmfb={os.path.join(model_test_dir, 'model.vmfb')}",
+            f"--parameters={os.path.join(model_test_dir, 'open-llama-3b-v2-f16.gguf')}",
+            f"--device={settings['device']}",
+        ]
+    )
+    # Wait for server to start
+    time.sleep(2)
+    yield server_process
+    # Teardown: kill the server
+    server_process.terminate()
+    server_process.wait()
+
+
+def do_generate(prompt):
+    headers = {"Content-Type": "application/json"}
+    # Create a GenerateReqInput-like structure
+    data = {
+        "text": prompt,
+        "sampling_params": {"max_tokens": 50, "temperature": 0.7},
+        "rid": uuid.uuid4().hex,
+        "return_logprob": False,
+        "logprob_start_len": -1,
+        "top_logprobs_num": 0,
+        "return_text_in_logprobs": False,
+        "stream": False,
+    }
+    print("Prompt text:")
+    print(data["text"])
+    BASE_URL = "http://localhost:8000"
+    response = requests.post(f"{BASE_URL}/generate", headers=headers, json=data)
```

Comment:
Is this port 8000 configurable? What happens if the port is in use? Will the test fail with a good error message?

Reply:
That's a good catch. Maybe a good route could be:

Before server starts we use `socket` to attempt to bind to the default port (8000). If that's unsuccessful, we increment the port until it finds an available one. Once we find an available port, we use that to start the server and proceed through the rest of the tests. Should probably set a limit on how many increments is allowed before failing, in case there's ever an error with a runner host that affects ability to bind to ANY ports.

We still add a good error message, just in case in the window between verifying a port's availability and starting the server, that port becomes unavailable.

Reply:
Created an `available_port` fixture to find an available dynamically, instead of having one hardcoded